src/tactic/rewrite_search/explain.lean

```diff
@@ -0,0 +1,255 @@
+private meta def explain_tree (rs : list (expr × bool)) (tree : app_addr) :
+tactic (list string) :=
+list.join <$> option.to_list <$> explain_tree_aux rs tree
+
+/-
+Gather all rewrites into trees, then generate a line of code for each tree.
+The return value has one `conv_x` tactic on each line.
+-/
+private meta def explanation_lines (rs : list (expr × bool)) (s : side) :
+option app_addr → list how → tactic (list string)
+| none [] := return []
+| (some tree) [] := do
+  tacs ← explain_tree rs tree,
+  return $ if tacs.length = 0 then []
+  else ["conv_" ++ s.to_xhs ++ " { " ++ string.intercalate ", " tacs ++ " }"]
+| tree (h :: rest) := do
+  (new_tree, rest_if_fail) ← match h.addr with
+  | (some addr) := do
+    new_tree ← splice_in tree [h.rule_index] addr,
+    return (some new_tree, list.cons h rest)
+  | none := do
+    return (none, rest)
+  end,
+  match new_tree with
+  | some (new new_tree) := explanation_lines new_tree rest
+  | _ := do
+    line ← explanation_lines tree [],
+    lines ← explanation_lines none rest_if_fail,
+    return $ line ++ lines
+  end
+
+/-- Explain a list of rewrites using `conv_x` tactics. -/
+meta def explain_rewrites (rs : list (expr × bool)) (s : side) (hows : list how) :
+tactic string :=
+  string.intercalate ",\n" <$> explanation_lines rs s none hows
+
+end using_conv
+
+private meta def explain_rewrites_concisely (steps : list (expr × bool)) (needs_refl : bool) :
+tactic string := do
+  rules ← string.intercalate ", " <$> steps.mmap pp_rule,
+  return $ "erw [" ++ rules ++ "]" ++ (if needs_refl then ", refl" else "")
+
+/-
+Fails if we can't just use rewrite.
+Otherwise, returns 'tt' if we need a `refl` at the end.
+-/
+private meta def check_if_simple_rewrite_succeeds (rewrites : list (expr × bool)) (goal : expr) :
+tactic bool :=
+lock_tactic_state $ do
+  m ← mk_meta_var goal,
+  set_goals [m],
+  rewrites.mmap' $ λ q, rewrite_target q.1 {symm := q.2, md := semireducible},
+  (reflexivity reducible >> return ff) <|> (reflexivity >> return tt)
+
+/-- Construct a list of rewrites from a proof unit. -/
+meta def proof_unit.rewrites (u : proof_unit) (rs : list (expr × bool)) : list (expr × bool) :=
+  u.steps.filter_map $ how.to_rewrite rs
+
+/-- Construct an explanation string from a proof unit. -/
+meta def proof_unit.explain (u : proof_unit) (rs : list (expr × bool))
+(explain_using_conv : bool) : tactic string :=
+  if explain_using_conv then
+    using_conv.explain_rewrites rs u.side u.steps
+  else
+    using_location.explain_rewrites rs u.side u.steps
+
+private meta def explain_proof_full (rs : list (expr × bool)) (explain_using_conv : bool) :
+list proof_unit → tactic string
+| [] := return ""
+| (u :: rest) := do
+  -- Don't use transitivity for the last unit, since it must be redundant.
+  head ← if rest.length = 0 ∨ u.side = side.L then pure [] else (do
+    n ← infer_type u.proof >>= (λ e, prod.snd <$> split_equation e) >>= pp,
+    pure $ ["transitivity " ++ to_string n]
+  ),
+
```

Comment:
Is there a reason you would like to keep this empty line?
```suggestion
```

Reply:
I don't mind removing it. done